pkg/storage/redis/redis.go

```diff
@@ -208,6 +210,60 @@ func (db *DB) ClearDeleted(_ context.Context, cutoff time.Time) {
 	}
 }
 
+func doNotify(ctx context.Context, psc *redis.PubSubConn, ch chan struct{}) {
+	switch v := psc.ReceiveWithTimeout(time.Second).(type) {
+	case redis.Message:
+		log.Debug().Str("action", string(v.Data)).Msg("Got redis message")
+		if string(v.Data) != watchAction {
+			return
+		}
+		select {
+		case <-ctx.Done():
+			log.Warn().Err(ctx.Err()).Msg("unable to notify channel")
+			return
+		case ch <- struct{}{}:
+		}
+	case error:
+		log.Debug().Err(v).Msg("redis subscribe error")
```

Comment:
I think we need to distinguish between timeout errors and other errors. Otherwise we'll just loop forever.

Reply:
I'm not sure how can we distinguish them. Also, isn't it ok to just loop forever? The loop is controlled by the context for termination.

Reply:
If, for example, you get disconnected from redis, we will end up calling this function, failing and then immediately trying again. I imagine this will peg the CPU at 100% on the process.

If the error is due to something internal to redis (not sure what could cause that) we might cause a denial of service attack due to all the failed requests.

Typically for error conditions we should use exponential backoff to prevent thundering herds like this.

I'm not sure what error redis returns on timeout, but if it's a net error you can use a method call to determine if the error was due to a timeout:

```
type Error interface {
	error
	Timeout() bool   // Is the error a timeout?
	Temporary() bool // Is the error temporary?
}
```

Reply:
Fixed.